zipextimporter.py

```diff
@@ -42,138 +43,227 @@
 
 """
 import sys
-import zipimport
-
-from importlib import machinery
-from importlib import util
+import _imp
+from zipimport import zipimporter
+try:
+    import _warnings as warnings
+except ImportError:
+    import warnings
+try:
+    from _frozen_importlib import ModuleSpec, spec_from_loader
+except ImportError:
+    from importlib.machinery import ModuleSpec
+    from importlib.util import spec_from_loader
 
 # _memimporter is a module built into the py2exe runstubs.
 import _memimporter
 
-class ZipExtensionImporter(zipimport.zipimporter):
-    _suffixes = machinery.EXTENSION_SUFFIXES
-
-    def find_loader(self, fullname, path=None):
-        """We need to override this method for Python 3.x.
-        """
-        loader, portions = super().find_loader(fullname, path)
-        if loader is None:
-            pathname = fullname.replace(".", "\\")
-            for s in self._suffixes:
-                if (pathname + s) in self._files:
-                    return self, []
-            return None, []
-        return loader, portions
-
-    def find_module(self, fullname, path=None):
-        result = zipimport.zipimporter.find_module(self, fullname, path)
-        if result:
-            return result
-        fullname = fullname.replace(".", "\\")
-        for s in self._suffixes:
-            if (fullname + s) in self._files:
-                return self
-        return None
-
-    def find_spec(self, name, path=None):
-        module = self.find_module(name, path)
-        if module is not None:
-            return util.spec_from_loader(name, module)
+
+class _ModuleInfo:
+    __slots__ = ("path", "is_ext", "is_package")
+    def __init__(self, *args):
+        self.path, self.is_ext, self.is_package = args
+
+
+class ZipExtensionImporter(zipimporter):
+    suffixes = _imp.extension_suffixes()
+    _searchorder = (
+        *[(f"\\__init__{suffix}", True, True) for suffix in suffixes],
+        ("\\__init__.pyc", False, True),
+        ("\\__init__.py", False, True),
+        *[(suffix, True, False) for suffix in suffixes],
+        (".pyc", False, False),
+        (".py", False, False),
+    )
+    _searchorder_pywin32 = (
+        *_searchorder[:-2],
+        ("%d%d.dll" % sys.version_info[:2], True, False),
+        *_searchorder[-2:],
+    )
+    del suffixes
```

Comment:
I am not a fan of `del` so I would use it only if it is absolutely necessary.

Reply:
Agree with you, this is not necessary.